File: protocol-generator/typescript/src/model/entity.ts
import { encode, decode } from "../codec/canonical-cbor.js";
import { encodeEntity } from "../codec/entity-codec.js";
import { type EcfValue, ecfBytes, ecfMap, ecfPreEncoded, ecfText } from "../codec/ecf-value.js";
import { SHA256_FORMAT, contentHashForFormat, readFormatCode } from "../codec/hash-formats.js";
import { EntityProtocolError } from "../errors.js";
import * as Ecf from "./ecf.js";
import { hashEqual, hashHex } from "./hashes.js";

/**
 * A materialized entity — `{type, data, content_hash}` (V7 §1.1, §3.4). The
 * fundamental data unit: a typed payload with content-addressed identity.
 *
 * Entity fidelity (§1.8) is load-bearing. An entity decoded from the wire retains
 * its exact original bytes in {@link wireBytes}; forwarding re-emits those bytes
 * verbatim and never re-serializes the decoded structure. The content hash is
 * computed once over `{type, data}` and trusted thereafter.
 */
export class Entity {
  private constructor(
    /** Semantic type path, e.g. `"system/protocol/execute"`. */
    readonly type: string,
    /** Decoded typed payload. */
    readonly data: EcfValue,
    /** 33-byte content hash (format code + SHA-256 digest), validated. */
    readonly contentHash: Uint8Array,
    /**
     * The exact `{type, data, content_hash}` wire bytes. For a locally-built
     * entity these are the canonical encoding; for a decoded entity they are the
     * original received bytes (§1.8 forward-original).
     */
    readonly wireBytes: Uint8Array,
  ) {}

  get contentHashHex(): string {
    return hashHex(this.contentHash);
  }

  /**
   * Build an entity from `{type, data}`: canonical-encode the hashable form,
   * derive the content hash under `contentHashFormat` (default `0x00` SHA-256 —
   * the §9.1 home format), and produce the full wire bytes. A non-default format
   * is the agility path (e.g. a SHA-384 home network).
   */
  static create(type: string, data: EcfValue, contentHashFormat: bigint = SHA256_FORMAT): Entity {
    const canonicalData = encode(data);
    const hashable = encodeEntity(type, canonicalData);
    const contentHash = contentHashForFormat(contentHashFormat, hashable);
    const wire = encode(
      ecfMap([
        [ecfText("type"), ecfText(type)],
        [ecfText("data"), ecfPreEncoded(canonicalData)],
        [ecfText("content_hash"), ecfBytes(contentHash)],
      ]),
    );
    return new Entity(type, data, contentHash, wire);
  }

  /**
   * Decode a wire entity and validate its content hash on receipt (§1.8.1,
   * §7.2). Throws {@link EntityProtocolError} on a hash mismatch or a malformed
   * shape.
   */
  static decode(wireBytes: Uint8Array): Entity {
    const value = decode(wireBytes);
    const type = Ecf.requireText(value, "type");
    const data = Ecf.require(value, "data");
    const declared = Ecf.requireBytes(value, "content_hash");

    // Wire-acceptance carve-out (§7.65): recompute under the format the entity
    // declares, so an entity authored under any supported home format validates.
    const format = readFormatCode(declared);
    const hashable = encodeEntity(type, encode(data));
    const expected = contentHashForFormat(format, hashable);
    if (!hashEqual(expected, declared)) {
      throw new EntityProtocolError(
        `content_hash mismatch on '${type}': computed ${hashHex(expected)}, declared ${hashHex(declared)}`,
      );
    }
    return new Entity(type, data, declared, wireBytes);
  }

  /**
   * Build an entity from an already-decoded envelope sub-value. The envelope is
   * decoded in strict canonical mode, so re-encoding this sub-map reproduces its
   * exact original bytes (ECF canonical form is unique) — non-canonical input was
   * rejected at the envelope boundary. This recovers the per-entity byte
   * boundaries the whole-envelope decode flattened, then validates the hash.
   */
  static fromDecoded(entityMap: EcfValue): Entity {
    return Entity.decode(encode(entityMap));
  }
}

File: protocol-generator/csharp/src/EntityCore.Protocol/Model/Entity.cs
using EntityCore.Protocol.Codec;

namespace EntityCore.Protocol.Model;

/// <summary>
/// A materialized entity — <c>{type, data, content_hash}</c> (V7 §1.1, §3.4). The
/// fundamental data unit: a typed payload with content-addressed identity.
/// <para>
/// Entity fidelity (§1.8) is load-bearing. An entity decoded from the wire retains
/// its exact original bytes in <see cref="WireBytes"/>; forwarding re-emits those
/// bytes verbatim and never re-serializes the decoded structure. The content hash
/// is computed once over <c>{type, data}</c> and trusted thereafter.
/// </para>
/// </summary>
internal sealed class Entity
{
    private Entity(string type, EcfValue data, byte[] contentHash, ReadOnlyMemory<byte> wireBytes)
    {
        Type = type;
        Data = data;
        ContentHash = contentHash;
        WireBytes = wireBytes;
    }

    /// <summary>Semantic type path, e.g. <c>"system/protocol/execute"</c>.</summary>
    public string Type { get; }

    /// <summary>Decoded typed payload.</summary>
    public EcfValue Data { get; }

    /// <summary>33-byte content hash (format code + SHA-256 digest), validated.</summary>
    public byte[] ContentHash { get; }

    /// <summary>
    /// The exact <c>{type, data, content_hash}</c> wire bytes. For a locally-built
    /// entity these are the canonical encoding; for a decoded entity they are the
    /// original received bytes (§1.8 forward-original).
    /// </summary>
    public ReadOnlyMemory<byte> WireBytes { get; }

    public string ContentHashHex => Hashes.Hex(ContentHash);

    /// <summary>
    /// The one entity type pinned to the ECFv1-SHA-256 floor regardless of the
    /// active or home <c>content_hash_format</c> (§4.5a item 1a).
    /// </summary>
    internal const string PeerIdentityType = "system/peer";

    /// <summary>
    /// Build an entity from <c>{type, data}</c>: canonical-encode the hashable form,
    /// derive the content hash under <paramref name="contentHashFormat"/> (default
    /// <c>0x00</c> SHA-256 — the §9.1 home format), and produce the full wire bytes.
    /// A non-default format is the agility path (e.g. a SHA-384 home network).
    /// </summary>
    /// <remarks>
    /// <para>
    /// §4.5a item 1a — a <c>system/peer</c> identity entity is authored under
    /// ECFv1-SHA-256 (<c>0x00</c>) <b>unconditionally</b>: on every connection,
    /// whatever the active format, and whatever the peer's home format. Its data is
    /// wholly recoverable from the public peer-id, so every consumer <i>derives</i>
    /// its hash rather than fetching it. A <c>system/peer</c> under any other format
    /// is not a form to be preserved but a construction that cannot exist, and
    /// authoring one is refused here.
    /// </para>
    /// <para>
    /// This is the constructor the <c>hash-format-sha-384.2</c> agility vector
    /// requires the refusal to be observed through. That vector used to assert the
    /// <i>opposite</i> — that the SHA-384 rehash succeeds — and stayed green only
    /// because the verifier hand-built the entity instead of routing through the
    /// code that forbids it. A fixture that exercises a forbidden construction and
    /// passes by bypassing the guard certifies the opposite of the rule
    /// (<c>GUIDE-CONFORMANCE</c> §2.4a).
    /// </para>
    /// <para>
    /// The guard is deliberately on the AUTHORING path only. <see cref="Decode"/>
    /// recomputes under the format an entity declares — that is wire acceptance, a
    /// separate surface, and item 1a does not speak to it.
    /// </para>
    /// </remarks>
    public static Entity Create(string type, EcfValue data, ulong contentHashFormat = HashFormats.Sha256)
    {
        if (type == PeerIdentityType && contentHashFormat != HashFormats.Sha256)
        {
            throw new EntityCodecException(
                $"'{PeerIdentityType}' is pinned to the ECFv1-SHA-256 floor (V7 section 4.5a item 1a); " +
                $"refusing to author it under content_hash_format 0x{contentHashFormat:x2}");
        }

        byte[] hashable = CanonicalCbor.Encode(Ecf.Map(("type", new EcfValue.Text(type)), ("data", data)));
        byte[] contentHash = HashFormats.ContentHash(contentHashFormat, hashable);
        byte[] wire = CanonicalCbor.Encode(new EcfValue.Map(new[]
        {
            new KeyValuePair<EcfValue, EcfValue>(new EcfValue.Text("type"), new EcfValue.Text(type)),
            new KeyValuePair<EcfValue, EcfValue>(new EcfValue.Text("data"), data),
            new KeyValuePair<EcfValue, EcfValue>(new EcfValue.Text("content_hash"), new EcfValue.Bytes(contentHash)),
        }));
        return new Entity(type, data, contentHash, wire);
    }

    /// <summary>
    /// Decode a wire entity and validate its content hash on receipt (§1.8.1,
    /// §7.2). Throws <see cref="EntityProtocolException"/> on a hash mismatch or a
    /// malformed shape.
    /// </summary>
    public static Entity Decode(ReadOnlyMemory<byte> wireBytes)
    {
        EcfValue value = CanonicalCbor.Decode(wireBytes);
        string type = Ecf.RequireText(value, "type");
        EcfValue data = Ecf.Require(value, "data");
        byte[] declared = Ecf.RequireBytes(value, "content_hash");

        // Wire-acceptance carve-out (§7.65): recompute under the format the entity
        // declares, so an entity authored under any supported home format validates.
        ulong format = HashFormats.ReadFormatCode(declared);
        byte[] hashable = CanonicalCbor.Encode(Ecf.Map(("type", new EcfValue.Text(type)), ("data", data)));
        byte[] expected = HashFormats.ContentHash(format, hashable);
        if (!Hashes.Equal(expected, declared))
        {
            // §1.8 item 1 resolution integrity, the same class as a mis-keyed `included`
            // entry and the same §5.2a code — `400 hash_mismatch`, never the structural
            // `invalid_request` beside it (0.8.2.24 N4/N5).
            throw new HashMismatchException(
                $"content_hash mismatch on '{type}': computed {Hashes.Hex(expected)}, declared {Hashes.Hex(declared)}");
        }

        return new Entity(type, data, declared, wireBytes);
    }

    /// <summary>
    /// Build an entity from an already-decoded envelope sub-value. The envelope is
    /// decoded in strict canonical mode, so re-encoding this sub-map reproduces its
    /// exact original bytes (ECF canonical form is unique) — non-canonical input
    /// was rejected at the envelope boundary. This recovers the per-entity byte
    /// boundaries the whole-envelope decode flattened, then validates the hash.
    /// </summary>
    public static Entity FromDecoded(EcfValue entityMap) => Decode(CanonicalCbor.Encode(entityMap));
}
